File: 01 - Leetcode/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        lpd = ['']

        def checkLpd(il: int, ir: int, lenPd: int):
            while il >= 0 and ir < n and s[il] == s[ir]:
                lenPd = ir - il + 1
                if lenPd > len(lpd[0]):
                    lpd[0] = s[il:ir+1]
                
                il-=1
                ir+=1

        for i in range(n): # O(n)
            # odd length
            checkLpd(i, i, 1)

            # even length
            checkLpd(i, i+1, 2)

        return lpd[0]

# @lc code=end

File: 01 - Leetcode/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
// https://www.youtube.com/watch?v=XYQecbcd6_c&ab_channel=NeetCode
class Solution
{
public:
    string lgp = "";
    string checkLpd(int il, int ir, int lenPd, string s)
    {
        int n = s.size();
        while (il >= 0 && ir < n && s[il] == s[ir]) // O(n)
        {
            lenPd = ir - il + 1;
            if (lenPd > lgp.size())
            {
                lgp = string(&s[il], &s[ir + 1]);
            }
            il--;
            ir++;
        }
        return lgp;
    }

    string longestPalindrome(string s)
    {
        int n = s.size();

        // O(n^2)
        for (int i = 0; i < n; i++) // O(n)
        {
            // odd length
            lgp = checkLpd(i, i, 1, s);

            // even length
            lgp = checkLpd(i, i + 1, 1, s);
        }
        return lgp;
    }
};
// @lc code=end
